tests/core/models/test_escn_compiles.py

```diff
@@ -0,0 +1,329 @@
+        mmax_list = [2],
+        sphere_channels = 128,
+        hidden_channels = 256,
+        edge_channels = 128,
+        num_sphere_samples = 128,
+        distance_function = "gaussian",
+        basis_width_scalar = 1.0,
+        distance_resolution = 0.02,
+        resolution = None,
+    )
+
+def init(backend: str):
+    if not torch.distributed.is_initialized():
+        init_local_distributed_process_group(backend=backend)
+
+class TestESCNCompiles:
+    def test_escn_baseline_cpu(self, tol=1e-8):
+        init("gloo")
+        data = load_data()
+        data_tg = data_list_collater([data])
+        data_export = data_list_collater([data], to_dict=True)
+
+        base_model = DistributedDataParallel(load_model("escn"))
+        export_model = DistributedDataParallel(load_model("escn_export"))
+        base_output = base_model(data_tg)
+        export_output = export_model(data_export)
+        torch.set_printoptions(precision=8)
+        assert torch.allclose(base_output["energy"], export_output["energy"], atol=tol)
+        assert torch.allclose(base_output["forces"].mean(0), export_output["forces"].mean(0), atol=tol)
+
+    @skip_if_no_cuda
+    def test_escn_baseline_cuda(self, tol=1e-8):
+        init("nccl")
+        data = load_data()
+        data_tg = data_list_collater([data]).to("cuda")
+        data_export = data_list_collater([data], to_dict=True)
+        data_export_cu = {k:v.to("cuda") for k,v in data_export.items()}
+
+        base_model = DistributedDataParallel(load_model("escn").cuda())
+        export_model = DistributedDataParallel(load_model("escn_export").cuda())
+        base_output = base_model(data_tg)
+        export_output = export_model(data_export_cu)
+        torch.set_printoptions(precision=8)
+        assert torch.allclose(base_output["energy"], export_output["energy"], atol=tol)
+        assert torch.allclose(base_output["forces"].mean(0), export_output["forces"].mean(0), atol=tol)
+
+    def test_rotation_invariance(self) -> None:
+        random.seed(1)
+        data = load_data()
+
+        # Sampling a random rotation within [-180, 180] for all axes.
+        transform = RandomRotate([-180, 180], [0, 1, 2])
+        data_rotated, rot, inv_rot = transform(data.clone())
+        assert not np.array_equal(data.pos, data_rotated.pos)
+
+        # Pass it through the model.
+        batch = data_list_collater([data, data_rotated], to_dict=True)
+        model = load_model("escn_export")
+        model.eval()
+        out = model(batch)
+
+        # Compare predicted energies and forces (after inv-rotation).
+        energies = out["energy"].detach()
+        np.testing.assert_almost_equal(energies[0], energies[1], decimal=5)
+
+        forces = out["forces"].detach()
+        logging.info(forces)
+        np.testing.assert_array_almost_equal(
+            forces[: forces.shape[0] // 2],
+            torch.matmul(forces[forces.shape[0] // 2 :], inv_rot),
+            decimal=5,
```

Comment:
with reasonable graphs (no max neighbor limit or no strict max neighbor) this should pass with 6 or even 7 decimal places.

Reply:
a single layer should get 7, seems like you are using 8 and that could be why.